File: backend/worker/security/crypto_manager.py
import os
import base64
from Crypto.Cipher import AES
from Crypto.Random import get_random_bytes

# The ALE_MASTER_KEY should be a 32-byte hex string (64 characters)
KEK_HEX = os.getenv("ALE_MASTER_KEY")

class CryptoManager:
    @staticmethod
    def decrypt_envelope(encrypted_data_bytes, encrypted_dek_b64, iv_b64, auth_tag_b64):
        """
        Decrypt data using envelope encryption.
        Matches the Node.js implementation logic.
        """
        if not KEK_HEX:
            raise ValueError("ALE_MASTER_KEY not configured in environment")
        
        kek = bytes.fromhex(KEK_HEX)
        
        # 1. Decode DEK and extract internal params
        full_encrypted_dek = base64.b64decode(encrypted_dek_b64)
        dek_iv = full_encrypted_dek[:12]
        dek_auth_tag = full_encrypted_dek[12:28]
        actual_encrypted_dek = full_encrypted_dek[28:]
        
        # 2. Decrypt DEK using KEK
        dek_cipher = AES.new(kek, AES.MODE_GCM, nonce=dek_iv)
        dek = dek_cipher.decrypt_and_verify(actual_encrypted_dek, dek_auth_tag)
        
        # 3. Decrypt data using decrypted DEK
        data_iv = base64.b64decode(iv_b64)
        data_auth_tag = base64.b64decode(auth_tag_b64)
        
        data_cipher = AES.new(dek, AES.MODE_GCM, nonce=data_iv)
        decrypted_data = data_cipher.decrypt_and_verify(encrypted_data_bytes, data_auth_tag)
        
        return decrypted_data

    @staticmethod
    def encrypt_envelope(data_bytes):
        """
        Encrypt data using envelope encryption.
        Matches the Node.js implementation logic.
        """
        if not KEK_HEX:
            raise ValueError("ALE_MASTER_KEY not configured in environment")
            
        kek = bytes.fromhex(KEK_HEX)
        
        # 1. Generate random DEK
        dek = get_random_bytes(32)
        
        # 2. Encrypt data with DEK
        data_iv = get_random_bytes(12)
        data_cipher = AES.new(dek, AES.MODE_GCM, nonce=data_iv)
        encrypted_data, data_auth_tag = data_cipher.encrypt_and_digest(data_bytes)
        
        # 3. Encrypt DEK with KEK
        dek_iv = get_random_bytes(12)
        dek_cipher = AES.new(kek, AES.MODE_GCM, nonce=dek_iv)
        encrypted_dek_body, dek_auth_tag = dek_cipher.encrypt_and_digest(dek)
        
        # Package DEK as [IV][TAG][ENCRYPTED_BODY]
        full_encrypted_dek = dek_iv + dek_auth_tag + encrypted_dek_body
        
        return {
            "encryptedData": encrypted_data,
            "encryptedDEK": base64.b64encode(full_encrypted_dek).decode('utf-8'),
            "iv": base64.b64encode(data_iv).decode('utf-8'),
            "authTag": base64.b64encode(data_auth_tag).decode('utf-8')
        }

File: backend/api/security/cryptoManager.js
/**
 * Crypto Manager
 * Handles Application-Level Encryption (ALE) using Envelope Encryption.
 * 
 * Strategy:
 * 1. Data is encrypted with a unique Data Encryption Key (DEK).
 * 2. The DEK is encrypted with a Master Key (KEK) from environment variables.
 * 3. AES-256-GCM is used for both encryption steps to ensure integrity.
 */

const crypto = require('crypto');

// The ALE_MASTER_KEY should be a 32-byte hex string (64 characters)
const KEK = process.env.ALE_MASTER_KEY;

if (!KEK || KEK.length !== 64) {
    console.error('[ALE] WARNING: ALE_MASTER_KEY is missing or invalid. Encryption will fail.');
}

/**
 * Encrypt a buffer using envelope encryption
 * @param {Buffer} data - The data to encrypt
 * @returns {Object} - { encryptedData, encryptedDEK, iv, authTag }
 */
function encryptEnvelope(data) {
    if (!KEK) throw new Error('ALE_MASTER_KEY not configured');

    // 1. Generate a random Data Encryption Key (DEK)
    const dek = crypto.randomBytes(32);

    // 2. Encrypt the data with the DEK
    const iv = crypto.randomBytes(12); // GCM standard IV length
    const cipher = crypto.createCipheriv('aes-256-gcm', dek, iv);

    const encryptedData = Buffer.concat([cipher.update(data), cipher.final()]);
    const authTag = cipher.getAuthTag();

    // 3. Encrypt the DEK with the Master Key (KEK)
    const kekBuffer = Buffer.from(KEK, 'hex');
    const dekIv = crypto.randomBytes(12);
    const dekCipher = crypto.createCipheriv('aes-256-gcm', kekBuffer, dekIv);

    const encryptedDEK = Buffer.concat([dekCipher.update(dek), dekCipher.final()]);
    const dekAuthTag = dekCipher.getAuthTag();

    return {
        encryptedData,
        // We store the encrypted DEK along with its own IV and auth tag
        encryptedDEK: Buffer.concat([dekIv, dekAuthTag, encryptedDEK]).toString('base64'),
        iv: iv.toString('base64'),
        authTag: authTag.toString('base64')
    };
}

/**
 * Decrypt a buffer using envelope encryption
 * @param {Buffer} encryptedData - The data to decrypt
 * @param {String} encryptedDEKBase64 - The DEK encrypted by KEK (includes IV and Tag)
 * @param {String} ivBase64 - The IV for the data
 * @param {String} authTagBase64 - The Auth Tag for the data
 * @returns {Buffer} - Decrypted data
 */
function decryptEnvelope(encryptedData, encryptedDEKBase64, ivBase64, authTagBase64) {
    if (!KEK) throw new Error('ALE_MASTER_KEY not configured');

    try {
        const kekBuffer = Buffer.from(KEK, 'hex');
        const fullEncryptedDEK = Buffer.from(encryptedDEKBase64, 'base64');

        // 1. Extract DEK encryption params
        const dekIv = fullEncryptedDEK.slice(0, 12);
        const dekAuthTag = fullEncryptedDEK.slice(12, 28);
        const actualEncryptedDEK = fullEncryptedDEK.slice(28);

        // 2. Decrypt the DEK using the KEK
        const dekDecipher = crypto.createDecipheriv('aes-256-gcm', kekBuffer, dekIv);
        dekDecipher.setAuthTag(dekAuthTag);
        const dek = Buffer.concat([dekDecipher.update(actualEncryptedDEK), dekDecipher.final()]);

        // 3. Decrypt the data using the decrypted DEK
        const iv = Buffer.from(ivBase64, 'base64');
        const authTag = Buffer.from(authTagBase64, 'base64');
        const decipher = crypto.createDecipheriv('aes-256-gcm', dek, iv);
        decipher.setAuthTag(authTag);

        return Buffer.concat([decipher.update(encryptedData), decipher.final()]);
    } catch (error) {
        console.error('[ALE] Decryption failed:', error.message);
        throw new Error('Data decryption failed. This usually means the key or integrity check is invalid.');
    }
}

module.exports = {
    encryptEnvelope,
    decryptEnvelope
};
